### PR TITLE
Add Rust and Kotlin CI checks

### LINKED ISSUE #3: Add small CI runs

### PR DESCRIPTION
**Pull request by tnull**
Closes #3.

This PR adds CIs for Rust and Kotlin, as well as licensing files for this project.

To this end we generate the Kotlin bindings and run some trivial tests to ensure the binding generation worked. 

This is mainly so we immediately see if we were to break binding generation going forward (and to enforce formatting, for example).

FWIW, the package names might be debatable (e.g., `org.rustbitcoin.bitcoin` vs `org.bitcoindevkit.bitcoinffi`), but for now I simply made them uniform so everything builds.

### CODE REVIEW

**Comment by reez** [comment]
Awesome good idea (CI and Licensing) 👍

Looking good to me, even though I'm not a Kotlin expert I was just reviewing that part by comparing to bdk-ffi mostly

File: uniffi-bindgen/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "uniffi-bindgen"
+authors = ["Elias Rohrer <dev@tnull.de>"]
+license = "MIT OR Apache-2.0"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+uniffi = { version = "0.28.0", features = ["cli"] }
```
[the remark this answers is not in the document]
Comment (thunderbiscuit) [reply uniffi-bindgen/Cargo.toml:9]: Ah interesting!

**Comment by thunderbiscuit** [comment]
I'm not against publishing the library of course, but I'd be tempted to wait until there is demand for it before adding all that plumbing to the codebase (especially that it implies engineering time/resources I'm not sure either of our projects has to spare at the moment). As is, I'm not yet sure why/how/who would use just a bunch of bitcoin types on their own without much logic (I'm sure there are use cases for it! but there are also great libraries in all of those languages that provide these primitives very well). The one exception to that might be the miniscript-related types, which are of course much more advanced in Rust than in any other languages. Still I'm tempted to wait until someone requests it.

As is, just as a good collaboration location for types many uniffi-based libraries will need (if we can be useful enough for others to start contributing their types as well), I suspect we will be able to grow the library to have a nice coverage of the rust-bitcoin ecosystem over the next year or so. From there if people need it on its own we could look at what languages exactly are required, who has the expertise to maintain, etc.? Just my 2 sats, I'm not actually against anything in this PR I just want to keep the conversation going.

**Comment by thunderbiscuit** [comment]
Yeah I mean I really didn't want to imply I'm against publishing it, but I was just thinking it'd be great to see if there is even any demand for these (currently) 10 types. The "plumbing" I was referring to is many-fold and ever changing with the evolving uniffi, but roughly:
1. Directories for each library
2. Each of those has build, test, release scripts
3. Tests
4. Packaging and publishing workflows
5. Accounts with the respective packaging directories and familiarity with their release (Maven Central, PyPI, the C# and Go ones, etc.).

To answer your specific questions:
1. > libraries that have undergone nearly as strict code review as rust-bitcoin?

That one is easy... 😆 probably none. (at least for miniscript stuff!)

2. > Great libraries in our target languages
 
That one is easy too. Note that many of those have seen much more "in-production-hours" than rust-bitcoin by maybe an order of magnitude. I'm not knocking on rust-bitcoin, I'm just saying we're not the only ones producing good bitcoin software. The need for good, high-level wallet and lightning software libraries is evident (to me at least) in this niche and for these languages, and is why bdk-ffi and ldk-node are such a good fit for the bindings. The need for a few low-level types in these languages... I don't see it as clearly yet.
    Java 👉 BitcoinJ, 
    Kotlin 👉 bitcoin-kmp (ACINQ)
    Python 👉 pycoin/bitcoinlib/bit
    Go 👉 The library maintained by Lightning Labs
    C# 👉 The library maintained by NIcolas Dorier for btcpayserver
    Swift 👉 I don't know this one enough to say

**Comment by thunderbiscuit** [comment]
In the meantime of course, anyone wanting these types without the domain-logic can still add `bdk-ffi` as a dependency, which will re-export (so far at least) all of the types in this library, and does so at no extra cost to the development team (we release anyway). This means you might add a few MB to your final binary but you're not out of options.

How do you feel about testing the waters with this library for a few months (say 6-8?) and seeing how things go and the library surface we cover before opening the box of releasing on all languages? At that point we can also discuss _which_ languages might be good targets, or whether all of them is just as easy once you do it for one.

**Comment by thunderbiscuit** [comment]
Ok one last thing: @tnull feel free to push back on this! If you think adding and maintaining the release of separate libraries for bitcoin-ffi is worth doing right now and are willing to add the code to the codebase I'm not against it at all and don't want to be a blocker for a good library to get out there.

**Review by thunderbiscuit — CHANGES_REQUESTED** [review]
Ok I took a final look and I think we can trim the Gradle scripts a bit to only use what we need. As is, the JVM part of it is only there for testing so it makes sense to have it lean and clean.

You're right I sort of confused 2 conversations together here. Merging this with related Rust and JVM tests is orthogonal to releasing the library on its own in the future. I made some comments, but overall it looks good to me.

File: uniffi-bindgen/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (thunderbiscuit) [reply uniffi-bindgen/Cargo.toml:9]: I _think_ if you pin your dependency but one of your dependencies is not pinned, on a new patch release of uniffi your `Cargo.lock` file and binary will end up with both versions in (0.28.0 and 0.28.1 say) unless you use the `[patch]` construct... is that right? Anyway I'm not 100% against it here, we'll probably just pin it in bdk-ffi if that's the case and I really don't want it to move. I certainly won't block merging on this small issue.

I'm also aware that it's very common practice in Rust (doesn't mean I like it!) to just let cargo find the latest patch of all your libraries and use that regardless of the version specified in your Cargo.toml file (or rather, that's the default behaviour, to override the version as long as patches have been released according to semver, unless you explicitly pin). So yeah it's probably what others expect too, and most project will probably have the uniffi dependency declared _not_ pinned; hence using the latest patch might work best for those in general.

**Review by thunderbiscuit — APPROVED** [review]
ACK ca2ab10964b80502a93aa168a1893d6e412c7535.

**Comment by Davidson-Souza** [comment]
super nit, but [9d73f5a](https://github.com/bitcoindevkit/bitcoin-ffi/pull/15/commits/9d73f5a113da02c97c2b7d437c8326e0715779b9) appears to have a loose `f` in the commit title.

Apart from that, ACK [ca2ab10](https://github.com/bitcoindevkit/bitcoin-ffi/pull/15/commits/ca2ab10964b80502a93aa168a1893d6e412c7535)

**Pull request merged**